### ISSUE TITLE
Question about sample from dataloader

### ISSUE THREAD

**Issue by playma** [facebookresearch/fairseq-py#40]
There are many places like the code below.
Can I only get partial data by itr or there is any good method? 

I want to sample just 40% of all data for training, and get data batch by batch.

Thank you.

    itr = dataset.dataloader(args.train_subset, num_workers=args.workers,
                        max_tokens=args.max_tokens, seed=args.seed, epoch=epoch,
                        max_positions=args.max_positions,
                        sample_without_replacement=args.sample_without_replacement,
                        skip_invalid_size_inputs_valid_test=args.skip_invalid_size_inputs_valid_test)

    num_sentences = 0

    with progress_bar(itr, smoothing=0, leave=False) as t:
        wps_meter = TimeMeter()
        gen_timer = StopwatchMeter()

**Comment by edunov** [comment]
Not sure I understand your question, do you want to sample 40% of training data and then use that sample to train for multiple epochs or do you want each epoch to be 40% of the data, but then every epoch you sample new 40%?  If the later, do you want to sample with replacement or without replacement? 

The first option can be easily added here: https://github.com/facebookresearch/fairseq-py/blob/master/fairseq/data.py#L279  justs sample from indices, e.g. 
```
indices = np.random.choice(indices, int(len(src) * 0.4), replace=False)
```

The second option is already implemented --sample-without-replacement  we don't provide sample with replacement because it performs significantly worse (your model will see less data)
